File: test/crypto.test.ts
/**
 * Unit tests for `@etzhayyim/sdk/crypto`.
 *
 * Scope: AEAD round-trip, AAD binding, tamper detection, key-id derivation.
 * Out of scope: PDS write/read integration (covered when encryptedWrite/
 * encryptedRead land on the main Etzhayyim class).
 */

import {describe, it, expect} from "vitest";
import {
  AEAD_ALG,
  AEAD_TAG_BYTES,
  ENVELOPE_VERSION,
  KEY_BYTES,
  NONCE_BYTES,
  PAD_BUCKETS,
  PAD_SCHEME_ISO7816,
  decrypt,
  encrypt,
  generateKey,
  generateNonce,
  keyIdOf,
  padIso7816,
  pickBucket,
  unpadIso7816,
} from "../src/crypto.js";

const SENDER = "did:web:alice.example";

describe("crypto envelope round-trip", () => {
  it("encrypts and decrypts a CBOR-serializable plaintext", () => {
    const key = generateKey();
    const plaintext = {hello: "world", n: 42, list: [1, 2, 3]};

    const env = encrypt({key, sender: SENDER, plaintext});

    expect(env.v).toBe(ENVELOPE_VERSION);
    expect(env.alg).toBe(AEAD_ALG);
    expect(env.nonce).toHaveLength(NONCE_BYTES);
    expect(env.sender).toBe(SENDER);
    expect(env.keyId).toBe(keyIdOf(key));

    const out = decrypt<typeof plaintext>({key, envelope: env});
    expect(out).toEqual(plaintext);
  });

  it("propagates innerType when supplied", () => {
    const key = generateKey();
    const env = encrypt({
      key,
      sender: SENDER,
      plaintext: {x: 1},
      innerType: "com.etzhayyim.governance.proposal",
    });
    expect(env.innerType).toBe("com.etzhayyim.governance.proposal");
  });

  it("uses a deterministic nonce when caller supplies one", () => {
    const key = generateKey();
    const nonce = generateNonce();
    const env = encrypt({key, sender: SENDER, plaintext: {a: 1}, nonce});
    expect(env.nonce).toEqual(nonce);
  });
});

describe("crypto AEAD binding", () => {
  it("rejects ciphertext when AAD differs at decrypt", () => {
    const key = generateKey();
    const env = encrypt({
      key,
      sender: SENDER,
      plaintext: {confidential: true},
      aad: new Uint8Array([1, 2, 3]),
    });
    expect(() =>
      decrypt({key, envelope: env, aad: new Uint8Array([9, 9, 9])})
    ).toThrow();
  });

  it("rejects ciphertext when AAD is missing at decrypt", () => {
    const key = generateKey();
    const env = encrypt({
      key,
      sender: SENDER,
      plaintext: {confidential: true},
      aad: new Uint8Array([1, 2, 3]),
    });
    expect(() => decrypt({key, envelope: env})).toThrow();
  });

  it("detects ciphertext tampering", () => {
    const key = generateKey();
    const env = encrypt({key, sender: SENDER, plaintext: {x: 1}});
    env.ciphertext[0] ^= 0xff;
    expect(() => decrypt({key, envelope: env})).toThrow();
  });

  it("rejects wrong key", () => {
    const key1 = generateKey();
    const key2 = generateKey();
    const env = encrypt({key: key1, sender: SENDER, plaintext: {x: 1}});
    expect(() => decrypt({key: key2, envelope: env})).toThrow();
  });
});

describe("crypto invariants", () => {
  it("rejects key of wrong length", () => {
    const badKey = new Uint8Array(16) as ReturnType<typeof generateKey>;
    expect(() => encrypt({key: badKey, sender: SENDER, plaintext: {}})).toThrow();
  });

  it("rejects nonce of wrong length", () => {
    const key = generateKey();
    const badNonce = new Uint8Array(12);
    expect(() =>
      encrypt({key, sender: SENDER, plaintext: {}, nonce: badNonce})
    ).toThrow();
  });

  it("rejects unknown envelope version at decrypt", () => {
    const key = generateKey();
    const env = encrypt({key, sender: SENDER, plaintext: {}});
    const bumped = {...env, v: 99 as unknown as typeof ENVELOPE_VERSION};
    expect(() => decrypt({key, envelope: bumped})).toThrow(/envelope version/);
  });

  it("keyIdOf is deterministic and 16 hex chars", () => {
    const key = generateKey();
    const id1 = keyIdOf(key);
    const id2 = keyIdOf(key);
    expect(id1).toBe(id2);
    expect(id1).toHaveLength(16);
    expect(id1).toMatch(/^[0-9a-f]+$/);
  });

  it("generateKey returns 32 bytes", () => {
    expect(generateKey()).toHaveLength(KEY_BYTES);
  });
});

describe("crypto padding (ADR-2605181200)", () => {
  it("padIso7816 appends 0x80 then 0x00s; unpad recovers original", () => {
    const plain = new Uint8Array([1, 2, 3, 4, 5]);
    const padded = padIso7816(plain, 16);
    expect(padded.length).toBe(16);
    expect(padded[5]).toBe(0x80);
    for (let i = 6; i < 16; i++) expect(padded[i]).toBe(0);
    expect(Array.from(unpadIso7816(padded))).toEqual(Array.from(plain));
  });

  it("padIso7816 rejects target too small", () => {
    expect(() => padIso7816(new Uint8Array(10), 10)).toThrow(/too small/);
  });

  it("unpadIso7816 rejects missing delimiter", () => {
    expect(() => unpadIso7816(new Uint8Array(16))).toThrow(/invalid/);
  });

  it("pickBucket picks the smallest bucket that fits plaintext + delimiter + tag", () => {
    // 1 byte plaintext + 1 delimiter + 16 tag = 18 → bucket 1024
    expect(pickBucket(1)).toBe(1024);
    // 1007 + 1 + 16 = 1024 → fits
    expect(pickBucket(1024 - 1 - AEAD_TAG_BYTES)).toBe(1024);
    // 1008 + 1 + 16 = 1025 → next bucket
    expect(pickBucket(1024 - AEAD_TAG_BYTES)).toBe(4096);
    // largest bucket
    expect(pickBucket(60000)).toBe(65536);
  });

  it("pickBucket throws when plaintext exceeds largest bucket", () => {
    expect(() => pickBucket(100000)).toThrow(/exceeds/);
  });

  it("encrypt(pad: 'bucket') yields ciphertext exactly equal to the bucket size", () => {
    const key = generateKey();
    const env = encrypt({key, sender: "did:test:a", plaintext: {x: 1}, pad: "bucket"});
    expect(env.pad).toBe(PAD_SCHEME_ISO7816);
    expect(env.ciphertext.length).toBe(PAD_BUCKETS[0]); // 1024 — tiny plaintext
    const back = decrypt({key, envelope: env});
    expect(back).toEqual({x: 1});
  });

  it("encrypt(pad: {bucket: 4096}) forces the explicit bucket", () => {
    const key = generateKey();
    const env = encrypt({
      key,
      sender: "did:test:a",
      plaintext: {x: 1},
      pad: {bucket: 4096},
    });
    expect(env.ciphertext.length).toBe(4096);
    const back = decrypt({key, envelope: env});
    expect(back).toEqual({x: 1});
  });

  it("differently-sized plaintexts encrypt to the same bucket length", () => {
    const key = generateKey();
    const small = encrypt({key, sender: "did:test:a", plaintext: {x: 1}, pad: "bucket"});
    const bigger = encrypt({
      key,
      sender: "did:test:a",
      plaintext: {body: "x".repeat(500), tag: "fill"},
      pad: "bucket",
    });
    expect(small.ciphertext.length).toBe(bigger.ciphertext.length);
    expect(small.ciphertext.length).toBe(PAD_BUCKETS[0]);
  });

  it("pad: 'none' (default) leaves ciphertext at native CBOR size + tag", () => {
    const key = generateKey();
    const env = encrypt({key, sender: "did:test:a", plaintext: {x: 1}});
    expect(env.pad).toBeUndefined();
    // Native CBOR of {x: 1} is 4 bytes; AEAD adds 16-byte tag → 20.
    expect(env.ciphertext.length).toBeLessThan(64);
  });

  it("envelope with pad set decrypts back to plaintext (no caller-side unpad arg)", () => {
    const key = generateKey();
    const obj = {body: "hi", n: 7};
    const env = encrypt({key, sender: "did:test:a", plaintext: obj, pad: "bucket"});
    expect(decrypt({key, envelope: env})).toEqual(obj);
  });
});

File: test/kotoba/lang/pqh/crypto_test.clj
(ns kotoba.lang.pqh.crypto-test
  "Port of crypto.test.ts, plus cross-language known-answer vectors
   (kotoba.lang.pqh.vectors) proving this namespace's XChaCha20-Poly1305 AEAD
   and canonical CBOR encoding are byte-identical to @noble/ciphers'
   xchacha20poly1305 and @ipld/dag-cbor's encode -- not just \"does it run\"."
  (:require [clojure.test :refer [deftest is testing use-fixtures]]
            [cbor.core :as cbor]
            [kotoba.lang.pqh.crypto :as crypto]
            [kotoba.lang.pqh.aead-bc :as aead-bc]
            [kotoba.lang.pqh.util :as u]
            [kotoba.lang.pqh.vectors :as v]))

(def SENDER "did:web:alice.example")

(use-fixtures :each (fn [f] (binding [crypto/*aead* (aead-bc/bc-aead)] (f))))

;; ── cross-language known-answer vectors ──────────────────────────────────

(deftest xchacha20poly1305-matches-noble
  (doseq [k [:xchacha20poly1305 :xchacha20poly1305_empty :xchacha20poly1305_bin]]
    (testing (str k)
      (let [vec (get v/vectors k)
            key (v/hb (:key vec))
            nonce (v/hb (:nonce vec))
            aad (some-> (:aad vec) v/hb)
            plaintext (v/hb (or (:plaintext vec) ""))
            ciphertext (crypto/xchacha20poly1305-encrypt key nonce aad plaintext)]
        (is (= (:ciphertext vec) (u/bytes->hex ciphertext)))
        (is (= (seq plaintext) (seq (crypto/xchacha20poly1305-decrypt key nonce aad ciphertext))))))))

(deftest cbor-encode-matches-dag-cbor
  (is (= (:encoded (:cbor_body1 v/vectors))
         (u/bytes->hex (cbor/encode (array-map "hello" "world" "n" 42 "list" [1 2 3])))))
  (is (= (:encoded (:cbor_body2 v/vectors))
         (u/bytes->hex (cbor/encode (array-map "x" 1))))))

;; ── round-trip (port of "crypto envelope round-trip") ────────────────────

(deftest envelope-round-trip
  (testing "encrypts and decrypts a CBOR-serializable plaintext"
    (let [key (crypto/generate-key)
          plaintext (array-map "hello" "world" "n" 42 "list" [1 2 3])
          env (crypto/encrypt {:key key :sender SENDER :plaintext plaintext})]
      (is (= crypto/ENVELOPE-VERSION (:v env)))
      (is (= crypto/AEAD-ALG (:alg env)))
      (is (= crypto/NONCE-BYTES (alength ^bytes (:nonce env))))
      (is (= SENDER (:sender env)))
      (is (= (crypto/key-id-of key) (:key-id env)))
      (is (= plaintext (crypto/decrypt {:key key :envelope env})))))

  (testing "propagates inner-type when supplied"
    (let [key (crypto/generate-key)
          env (crypto/encrypt {:key key :sender SENDER :plaintext (array-map "x" 1)
                                :inner-type "com.etzhayyim.governance.proposal"})]
      (is (= "com.etzhayyim.governance.proposal" (:inner-type env)))))

  (testing "uses a deterministic nonce when caller supplies one"
    (let [key (crypto/generate-key)
          nonce (crypto/generate-nonce)
          env (crypto/encrypt {:key key :sender SENDER :plaintext (array-map "a" 1) :nonce nonce})]
      (is (= (seq nonce) (seq (:nonce env)))))))

;; ── AEAD binding (port of "crypto AEAD binding") ──────────────────────────

(deftest aead-binding
  (testing "rejects ciphertext when AAD differs at decrypt"
    (let [key (crypto/generate-key)
          env (crypto/encrypt {:key key :sender SENDER :plaintext (array-map "confidential" true)
                                :aad (byte-array [1 2 3])})]
      (is (thrown? Exception (crypto/decrypt {:key key :envelope env :aad (byte-array [9 9 9])})))))

  (testing "rejects ciphertext when AAD is missing at decrypt"
    (let [key (crypto/generate-key)
          env (crypto/encrypt {:key key :sender SENDER :plaintext (array-map "confidential" true)
                                :aad (byte-array [1 2 3])})]
      (is (thrown? Exception (crypto/decrypt {:key key :envelope env})))))

  (testing "detects ciphertext tampering"
    (let [key (crypto/generate-key)
          env (crypto/encrypt {:key key :sender SENDER :plaintext (array-map "x" 1)})
          ct ^bytes (:ciphertext env)]
      (aset-byte ct 0 (unchecked-byte (bit-xor (aget ct 0) (unchecked-byte 0xff))))
      (is (thrown? Exception (crypto/decrypt {:key key :envelope env})))))

  (testing "rejects wrong key"
    (let [key1 (crypto/generate-key)
          key2 (crypto/generate-key)
          env (crypto/encrypt {:key key1 :sender SENDER :plaintext (array-map "x" 1)})]
      (is (thrown? Exception (crypto/decrypt {:key key2 :envelope env}))))))

;; ── invariants ────────────────────────────────────────────────────────────

(deftest invariants
  (testing "rejects key of wrong length"
    (is (thrown? Exception
                 (crypto/encrypt {:key (byte-array 16) :sender SENDER :plaintext {}}))))

  (testing "rejects nonce of wrong length"
    (let [key (crypto/generate-key)]
      (is (thrown? Exception
                   (crypto/encrypt {:key key :sender SENDER :plaintext {} :nonce (byte-array 12)})))))

  (testing "rejects unknown envelope version at decrypt"
    (let [key (crypto/generate-key)
          env (crypto/encrypt {:key key :sender SENDER :plaintext {}})
          bumped (assoc env :v 99)]
      (is (thrown-with-msg? clojure.lang.ExceptionInfo #"envelope version"
                            (crypto/decrypt {:key key :envelope bumped})))))

  (testing "key-id-of is deterministic and 16 hex chars"
    (let [key (crypto/generate-key)
          id1 (crypto/key-id-of key)
          id2 (crypto/key-id-of key)]
      (is (= id1 id2))
      (is (= 16 (count id1)))
      (is (re-matches #"[0-9a-f]+" id1))))

  (testing "generate-key returns 32 bytes"
    (is (= crypto/KEY-BYTES (alength (crypto/generate-key))))))

;; ── padding (ADR-2605181200) ──────────────────────────────────────────────

(deftest padding
  (testing "pad-iso7816 appends 0x80 then 0x00s; unpad recovers original"
    (let [plain (byte-array [1 2 3 4 5])
          padded (crypto/pad-iso7816 plain 16)]
      (is (= 16 (alength padded)))
      (is (= (unchecked-byte 0x80) (aget padded 5)))
      (doseq [i (range 6 16)] (is (zero? (aget padded i))))
      (is (= (seq (crypto/unpad-iso7816 padded)) (seq plain)))))

  (testing "pad-iso7816 rejects target too small"
    (is (thrown-with-msg? clojure.lang.ExceptionInfo #"too small"
                          (crypto/pad-iso7816 (byte-array 10) 10))))

  (testing "unpad-iso7816 rejects missing delimiter"
    (is (thrown-with-msg? clojure.lang.ExceptionInfo #"invalid"
                          (crypto/unpad-iso7816 (byte-array 16)))))

  (testing "pick-bucket picks the smallest bucket that fits plaintext + delimiter + tag"
    (is (= 1024 (crypto/pick-bucket 1)))
    (is (= 1024 (crypto/pick-bucket (- 1024 1 crypto/AEAD-TAG-BYTES))))
    (is (= 4096 (crypto/pick-bucket (- 1024 crypto/AEAD-TAG-BYTES))))
    (is (= 65536 (crypto/pick-bucket 60000))))

  (testing "pick-bucket throws when plaintext exceeds largest bucket"
    (is (thrown-with-msg? clojure.lang.ExceptionInfo #"exceeds"
                          (crypto/pick-bucket 100000))))

  (testing "encrypt(:pad :bucket) yields ciphertext exactly equal to the bucket size"
    (let [key (crypto/generate-key)
          env (crypto/encrypt {:key key :sender "did:test:a" :plaintext (array-map "x" 1) :pad :bucket})]
      (is (= crypto/PAD-SCHEME-ISO7816 (:pad env)))
      (is (= (first crypto/PAD-BUCKETS) (alength ^bytes (:ciphertext env))))
      (is (= (array-map "x" 1) (crypto/decrypt {:key key :envelope env})))))

  (testing "encrypt(:pad {:bucket 4096}) forces the explicit bucket"
    (let [key (crypto/generate-key)
          env (crypto/encrypt {:key key :sender "did:test:a" :plaintext (array-map "x" 1) :pad {:bucket 4096}})]
      (is (= 4096 (alength ^bytes (:ciphertext env))))
      (is (= (array-map "x" 1) (crypto/decrypt {:key key :envelope env})))))

  (testing "differently-sized plaintexts encrypt to the same bucket length"
    (let [key (crypto/generate-key)
          small (crypto/encrypt {:key key :sender "did:test:a" :plaintext (array-map "x" 1) :pad :bucket})
          bigger (crypto/encrypt {:key key :sender "did:test:a"
                                   :plaintext (array-map "body" (apply str (repeat 500 "x")) "tag" "fill")
                                   :pad :bucket})]
      (is (= (alength ^bytes (:ciphertext small)) (alength ^bytes (:ciphertext bigger))))
      (is (= (first crypto/PAD-BUCKETS) (alength ^bytes (:ciphertext small))))))

  (testing "pad :none (default) leaves ciphertext at native CBOR size + tag"
    (let [key (crypto/generate-key)
          env (crypto/encrypt {:key key :sender "did:test:a" :plaintext (array-map "x" 1)})]
      (is (nil? (:pad env)))
      (is (< (alength ^bytes (:ciphertext env)) 64))))

  (testing "envelope with pad set decrypts back to plaintext (no caller-side unpad arg)"
    (let [key (crypto/generate-key)
          obj (array-map "body" "hi" "n" 7)
          env (crypto/encrypt {:key key :sender "did:test:a" :plaintext obj :pad :bucket})]
      (is (= obj (crypto/decrypt {:key key :envelope env}))))))
